[torchao] safetensors

liangel-02 opened the pull request:
**Summary**
updating safetensors logic to support all tensor subclasses generically 

**Test**
verified integration with vllm 
unit tests pass: `python tests/quantization/torchao_integration/test_torchao.py -k TorchAoSafeSerializationTest`

liangel-02 committed:
torchao safetensors

liangel-02 committed:
working

liangel-02 committed:
update torchao safetensors

liangel-02 committed:
after rebase

jerryzh168 commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -91,6 +92,11 @@ def _linear_extra_repr(self):
+        torchao.quantization.Int4WeightOnlyConfig,
+        torchao.quantization.IntxWeightOnlyConfig,
+        torchao.quantization.Int8DynamicActivationIntxWeightConfig,
+        torchao.quantization.ModuleFqnToConfig,
```
nit: maybe add `torchao.quantization.FqnToConfig` as well, that's going to replace `ModuleFqnToConfig` I think

Also after the PR, can we turn on safe tensor serialization by default for all non-prototype configs now, or how far are we from that?

MekkCyber commented on src/transformers/core_model_loading.py:
```diff
@@ -860,7 +887,15 @@ def convert_and_load_state_dict_in_model(
+                pattern_regex = re.compile(f"^{pattern_with_captures}$")
+                concrete_source_pattern = extract_concrete_key_from_regex_pattern(
+                    original_key, source_pattern, pattern_regex
+                )
```
why do we need this ? could you explain a bit more why we need to change the regex handling here and in the other parts of the code?

MekkCyber reviewed (commented):
Thanks a lot for working on this

liangel-02 committed:
remove regex

liangel-02 commented:
@MekkCyber to avoid adding torchao specific logic to the core model loader and unblock landing, i reverted the changes to the regex logic and added all other tensor data names. in the future, maybe we can implement a strategy that generalizes this matching in a better way?

cc @jerryzh168

jerryzh168 commented on src/transformers/integrations/torchao.py:
```diff
@@ -215,56 +215,46 @@ def convert(
+            and return it as a dictionary with the full_layer_name as the key and the recovered
+            Float8Tensor instance as the value.
+        """
+        is_unsafe_serialization = "_weight_" not in list(input_dict.keys())[0]
```
this is using the same logic, we can probably make this more robust as well

jerryzh168 commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -249,8 +244,6 @@ def param_needs_quantization(self, model: "PreTrainedModel", param_name: str, **
         # check if the param_name is not in self.modules_to_not_convert
         if any(key + "." in param_name or key == param_name for key in self.modules_to_not_convert):
             return False
-        elif any(param_name.endswith(f":{x}") for x in self.full_ao_keys):
```
for this one, maybe just change it to `(f"_{x} for x in self.full_ao_keys)"` to be safe, at least with this we are confident it's correct. if there are no better general way to detect safetensors

liangel-02 replied:
i think this is achieving the same thing as L237, i think we only need one check

liangel-02 committed:
make safetensors check more robust

liangel-02 commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -241,6 +234,8 @@ def update_unexpected_keys(self, model, unexpected_keys: list[str]) -> list[str]
 
     def param_needs_quantization(self, model: "PreTrainedModel", param_name: str, **kwargs) -> bool:
+        if "_weight_" in param_name:
+            return True
```
i think i can remove this actually

MekkCyber commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -548,15 +536,18 @@ def get_weight_conversions(self):
+                        "_weight_scale",
+                        "_weight_zero_point",
+                        "_weight_act_pre_scale",
+                    ],
```
Yes it is greedy but to match for example `_weight_scale` and not `_weight_scale_and_zero` you can just do something like `_weight_scale$`, but ordering the keys works as well 👍

MekkCyber commented on src/transformers/modeling_utils.py:
```diff
@@ -2104,7 +2104,6 @@ def set_decoder(self, decoder):
         for name in possible_module_names:
             if hasattr(self, name):
-                print(name)
                 setattr(self, name, decoder)
```
Thanks for catching this

MekkCyber commented on src/transformers/integrations/torchao.py:
```diff
@@ -210,61 +210,52 @@ def __init__(self, hf_quantizer):
-            else:
-                param_data[f"{full_layer_name}:qdata"] = input_dict["weight:qdata"]
+            for suffix in input_dict.keys():
+                assert len(input_dict[suffix]) == 1
```
let's do an `if/else` and raise an error to follow the same pattern we use in transformers

MekkCyber reviewed (approved):
Very nice alternative solution ! Thanks a lot

SunMarc commented on src/transformers/integrations/torchao.py:
```diff
@@ -210,61 +210,52 @@ def __init__(self, hf_quantizer):
+        is_unsafe_serialization = list(input_dict.keys())[0] not in source_patterns
 
         param_data = {}
+        layer_name = ".".join(full_layer_name.split(".")[:-1])
```
Let's use rsplit as we do for other methods 
```suggestion
        layer_name = full_layer_name.rsplit(".", 1)[0]
```

liangel-02 commented:
@MekkCyber @SunMarc can you help me merge this, thanks!

MekkCyber committed:
Merge branch 'main' into ao_safetensors

ArthurZucker reviewed (approved):
LGTM
